From 3cd71dfdb89050b17a5e5e7c80c3d8e0cd02b196 Mon Sep 17 00:00:00 2001
From: Srinivas Pandruvada <srinivas.pandruvada@intel.com>
From: Evgeni <evgeni.raikhel@intel.com>
Date: Sun, 13 May 2018 10:49:26 -0400

Subject: [PATCH] Adding missing HID timestamp patch for Gyro sensor.
         Ubuntu 20.04 Bionic Beaver with Kernel 5.8
	A symmetric patch for Accelerator was already upstreamed,
	this patch was skipped from upstream
	The patch was written by Srinivas Pandruvada <srinivas.pandruvada@intel.com>
	The version tag has been added to accel module so it will be also replaced regardless
	of the target kernel version

---
 drivers/iio/accel/hid-sensor-accel-3d.c |  2 +
 drivers/iio/gyro/hid-sensor-gyro-3d.c   | 31 ++++++++++++++++++-------
 2 files changed, 24 insertions(+), 8 deletions(-)

diff --git a/5.8.0/drivers/iio/accel/hid-sensor-accel-3d.c b/5.8.0/drivers/iio/accel/hid-sensor-accel-3d.c
index 41d97faf5..5f5c31a94 100644
--- a/5.8.0/drivers/iio/accel/hid-sensor-accel-3d.c
+++ b/5.8.0/drivers/iio/accel/hid-sensor-accel-3d.c
@@ -274,6 +274,7 @@ static int accel_3d_capture_sample(struct hid_sensor_hub_device *hsdev,
 			hid_sensor_convert_timestamp(
 					&accel_state->common_attributes,
 					*(int64_t *)raw_data);
+		ret = 0;
 	break;
 	default:
 		break;
@@ -467,3 +467,4 @@ module_platform_driver(hid_accel_3d_platform_driver);
 MODULE_DESCRIPTION("HID Sensor Accel 3D");
 MODULE_AUTHOR("Srinivas Pandruvada <srinivas.pandruvada@intel.com>");
 MODULE_LICENSE("GPL");
+MODULE_VERSION("realsense2-dkms");
diff --git a/5.8.0/drivers/iio/gyro/hid-sensor-gyro-3d.c b/5.8.0/drivers/iio/gyro/hid-sensor-gyro-3d.c
index c67ce2a..e722b8f 100644
--- a/5.8.0/drivers/iio/gyro/hid-sensor-gyro-3d.c
+++ b/5.8.0/drivers/iio/gyro/hid-sensor-gyro-3d.c
@@ -27,11 +27,13 @@ struct gyro_3d_state {
 	struct hid_sensor_hub_callbacks callbacks;
 	struct hid_sensor_common common_attributes;
 	struct hid_sensor_hub_attribute_info gyro[GYRO_3D_CHANNEL_MAX];
-	u32 gyro_val[GYRO_3D_CHANNEL_MAX];
+	/* Reserve for 3 channels + padding + timestamp */
+	u32 gyro_val[GYRO_3D_CHANNEL_MAX + 3];
 	int scale_pre_decml;
 	int scale_post_decml;
 	int scale_precision;
 	int value_offset;
+	int64_t timestamp;
 };
 
 static const u32 gyro_3d_addresses[GYRO_3D_CHANNEL_MAX] = {
@@ -72,7 +74,8 @@ static const struct iio_chan_spec gyro_3d_channels[] = {
 		BIT(IIO_CHAN_INFO_SAMP_FREQ) |
 		BIT(IIO_CHAN_INFO_HYSTERESIS),
 		.scan_index = CHANNEL_SCAN_INDEX_Z,
-	}
+	},
+	IIO_CHAN_SOFT_TIMESTAMP(3)
 };
 
 /* Adjust channel real bits based on report descriptor */
@@ -179,11 +182,11 @@ static const struct iio_info gyro_3d_info = {
 };
 
 /* Function to push data to buffer */
-static void hid_sensor_push_data(struct iio_dev *indio_dev, const void *data,
-	int len)
+static void hid_sensor_push_data(struct iio_dev *indio_dev, void *data,
+	int len, int64_t timestamp)
 {
 	dev_dbg(&indio_dev->dev, "hid_sensor_push_data\n");
-	iio_push_to_buffers(indio_dev, data);
+	iio_push_to_buffers_with_timestamp(indio_dev, data, timestamp);
 }
 
 /* Callback handler to send event after all samples are received and captured */
@@ -195,11 +198,16 @@ static int gyro_3d_proc_event(struct hid_sensor_hub_device *hsdev,
 	struct gyro_3d_state *gyro_state = iio_priv(indio_dev);
 
 	dev_dbg(&indio_dev->dev, "gyro_3d_proc_event\n");
-	if (atomic_read(&gyro_state->common_attributes.data_ready))
+	if (atomic_read(&gyro_state->common_attributes.data_ready)) {
+		if (!gyro_state->timestamp)
+			gyro_state->timestamp = iio_get_time_ns(indio_dev);
+
 		hid_sensor_push_data(indio_dev,
 				gyro_state->gyro_val,
-				sizeof(gyro_state->gyro_val));
-
+				sizeof(gyro_state->gyro_val),
+				gyro_state->timestamp);
+		gyro_state->timestamp = 0;
+	}
 	return 0;
 }
 
@@ -223,6 +231,13 @@ static int gyro_3d_capture_sample(struct hid_sensor_hub_device *hsdev,
 						*(u32 *)raw_data;
 		ret = 0;
 	break;
+	case HID_USAGE_SENSOR_TIME_TIMESTAMP:
+		gyro_state->timestamp =
+			hid_sensor_convert_timestamp(
+					&gyro_state->common_attributes,
+					*(int64_t *)raw_data);
+		ret = 0;
+	break;
 	default:
 		break;
 	}
@@ -398,3 +413,4 @@ module_platform_driver(hid_gyro_3d_platform_driver);
 MODULE_DESCRIPTION("HID Sensor Gyroscope 3D");
 MODULE_AUTHOR("Srinivas Pandruvada <srinivas.pandruvada@intel.com>");
 MODULE_LICENSE("GPL");
+MODULE_VERSION("realsense2-dkms");
-- 
2.17.1

